cocotb/triggers.py

```diff
@@ -673,13 +654,12 @@ def __repr__(self):
         )
 
 
-@decorators.coroutine
-def _wait_callback(trigger, callback):
+async def _wait_callback(trigger, callback):
     """
-    Wait for a trigger, and call `callback` with the outcome of the yield.
+    Wait for a trigger, and call `callback` with the outcome of the await.
     """
     try:
-        ret = outcomes.Value((yield trigger))
+        ret = outcomes.Value(await trigger)
```

Comment:
This is a behavior change, and means that `fork(_wait_callback()).kill()` now throws an exception inside the `trigger` itself if `trigger` is an async coroutine and not a `RunningTask`.

I'd prefer we not change that behavior in this patch. Leaving this function untouched would restore the old behavior, and we can come back to it after more though.

Reply:
This change is fine, it's the change to `AggregateWaitable` that allows users to pass any Awaitable.

Reply:
Change was reverted.

Reply:
Unfortunately this still changes the semantics of cancellation inside `Waitable` objects.

The thing to realize is that `yield trigger` translates to `await scheduler._trigger_from_any(trigger)`.

If you used that here, then everything would work as before, and you'd still be able to pass async coroutines directly to `First`.

I'm ok with us calling a private function here as long as we add a todo comment about moving it somewhere more sensible.

Reply:
Ahh, forgot about Waitables. Advise what you think the proper solution is.

Reply:
I recommend you:

1. Put back your simplification in `AggregateWaitable`
2. Pick between:
   * Not converting this routine to use `async`, leave it using `cocotb.coroutine`
   * Using `trigger = scheduler._trigger_from_any(trigger)` above line +662.

I'd lean towards you just picking the second option in 2, since that makes the diff smaller.

Reply:
So if you pass an async coroutine, it would schedule it and it would be on the user to know to kill it after the fact? Seems like we should leave the simplification off until we commit to changing the semantics as that seems antagonistic.

Reply:
I believe that already happens today if you pass an async function.

Reply:
I'll add some tests for the quirks of the current behavior in a new PR.

Reply:
You currently can't pass it an async coroutine at all. There is a type guard that matches on `RunningTask`, `Waitable`, and `Trigger`.

Reply:
Good catch. However, it's still what happens today if you pass a `cocotb.coroutine` in.

Reply:
You can kill those however, if you pass in an un-decorated async coroutine how do kill it after `First` fires?

Reply:
Ah, I see. I was thinking about situations like:
```python
await First(some_coroutine(), some_trigger)
```
where the caller has already thrown away their hope of killing the coroutine. What you're describing I think is
```python
some_task = some_coroutine()
await First(some_task, some_trigger)
# user can call some_task.kill()
```

Perhaps it is best to play it safe and continue to forbid passing async coroutines to `First` until a later patch, as you've done.

Reply:
I'd like to use the first case, but I feel like `First` should do the scheduling *and* the killing in the case of passing an undecorated async coroutine, and if you needed to have it continue to run after `First`, you would wrap it in a `fork`, or perhaps something like `asyncio.shield`?